feat: Adding support for netsim to be able to work with Tools adapter

Rosme opened the pull request:
This adds an internal callback to tell the tools adapter that the transport is initialized and ready so we can act on it and be aware of it.

## Testing and Documentation

- No tests have been added.

Commit:
Adding support for netsim to be able to work with Tools adapter

Rosme commented on com.unity.netcode.gameobjects/Runtime/Transports/UTP/UnityTransport.cs:
```diff
@@ -1351,7 +1362,7 @@ private void ConfigureSimulator()
 #endif
 
-#if UNITY_EDITOR || DEVELOPMENT_BUILD
+#if UNITY_EDITOR || DEVELOPMENT_BUILD || MULTIPLAYER_TOOLS_1_1_0
```
This is based on an assumption that this will be defined when MULTIPLAYER_TOOLS_1_1_0 is released which would contain NetSim. @becksebenius-unity @fluong6

becksebenius-unity replied:
This would need to be added to the assembly definition as a conditional compiler directive, it won't happen automatically

ian-m-unity replied:
Should this `||` only occur if `NetSim` is actually included in the build? It's possible to have a release build without NetSim included right?

ian-m-unity commented on com.unity.netcode.gameobjects/Runtime/Transports/UTP/UnityTransport.cs:
```diff
@@ -346,6 +346,10 @@ private struct PacketLossCache
         };
 
+        internal static event Action<UnityTransport> TransportInitialized;
+        internal static event Action<UnityTransport> TransportDisposed;
```
Should a reference to the Transport be included in the disposed message? Is it still valid at that point?

Rosme replied:
Yes, since we are using the Instance ID to track the adapter.

JesseOlmer replied:
Can you expand on what the value is of tracking dispose instead of shutdown? Does this need to be an event or can you use a weak reference?

To expand on Ian's concern and my other comment about internals vs public, at the point you have this being invoked the object is essentially "off limits" to other objects as far as best practices go - regardless of how this may be used right now this feels dangerous and dependent on tribal knowledge about the control flow. If all you need is the ID, why not pass that as the event parameter instead of the entire (mostly disposed) object?

Rosme replied:
There's more to it. We are using the whole UnityTransport to access the Network Driver. Arguably, we could provide the Instance ID and the Network Driver only as arguments. I am not against this change. As for the Disposed, we definitely only need the Instance ID and I can change it in the same manner if you'd prefer that.

simon-lemay-unity commented on com.unity.netcode.gameobjects/Runtime/Transports/UTP/UnityTransport.cs:
```diff
@@ -1324,12 +1332,15 @@ private void ConfigureSimulator()
                 packetDelayMs: DebugSimulator.PacketDelayMS,
                 packetJitterMs: DebugSimulator.PacketJitterMS,
-                packetDropPercentage: DebugSimulator.PacketDropRate,
-                randomSeed: DebugSimulatorRandomSeed
```
Any reason this was removed?

Rosme replied:
Mistake. Will put it back

ian-m-unity commented on com.unity.netcode.gameobjects/Runtime/Transports/UTP/UnityTransport.cs:
```diff
@@ -1373,7 +1384,7 @@ private void ConfigureSimulator()
 #endif
 
-#if UNITY_EDITOR || DEVELOPMENT_BUILD
+#if UNITY_EDITOR || DEVELOPMENT_BUILD || MULTIPLAYER_TOOLS_1_1_0
```
Should this `||` only occur if `NetSim` is actually included in the build? It's possible to have a release build without NetSim included right?

Rosme replied:
You are right. Which means we will need the boilerplate stuff, urgh.

Commit:
Adding the simulator only if active in build inside unity transport.
 + adding back missing parameters

becksebenius-unity commented on com.unity.netcode.gameobjects/Runtime/Transports/UTP/UnityTransport.cs:
```diff
@@ -1330,6 +1347,10 @@ private void ConfigureSimulator()
 #endif
             );
+
+#if UTP_TRANSPORT_2_0_ABOVE
```
Should this PR not introduce this compiler directive? Or is there another tracking ticket for when it will be introduced?

Rosme replied:
This compile directive is already there through the existence of UTP 2. It was added by NGO team to support UTP 2.

becksebenius-unity replied:
Ah okay, thanks for confirming

Commit:
Setting up Simulator Pipeline stage if NetSim is available

JesseOlmer commented on com.unity.netcode.gameobjects/Runtime/Transports/UTP/UnityTransport.cs:
```diff
@@ -1348,12 +1369,12 @@ private void ConfigureSimulator()
             NetworkPipelineStageCollection.RegisterPipelineStage(new NetworkMetricsPipelineStage());
-#endif
-#endif
+#endif //!UTP_TRANSPORT_2_0_ABOVE
```
If you're going to be modifying these lines already why not just make this into 1 compound conditional?

JesseOlmer commented on com.unity.netcode.gameobjects/Runtime/Transports/UTP/UnityTransport.cs:
```diff
@@ -1362,76 +1383,81 @@ private void ConfigureSimulator()
             driver.RegisterPipelineStage<NetworkMetricsPipelineStage>(new NetworkMetricsPipelineStage());
-#endif
-#endif
+#endif //UTP_TRANSPORT_2_0_ABOVE
```
Again, either don't make changes to code you're not changing (this is the preferred option), or if you're going to make changes let's update the code with a compound conditional rather than nested preprocessor instructions

JesseOlmer replied:
To be clear: All of these #endif comment modifications interlaced with actual functional changes makes the PR much more difficult to review - next time separate the two matters, please.

Rosme replied:
I can revert those changes if you wish. I found it easier to follow that way.

JesseOlmer commented on com.unity.netcode.gameobjects/Runtime/AssemblyInfo.cs:
```diff
@@ -12,3 +12,4 @@
 [assembly: InternalsVisibleTo("Unity.Netcode.RuntimeTests")]
 [assembly: InternalsVisibleTo("Unity.Netcode.TestHelpers.Runtime")]
 [assembly: InternalsVisibleTo("Unity.Netcode.Adapter.UTP")]
+[assembly: InternalsVisibleTo("Unity.Multiplayer.Tools.Adapters.Ngo1WithUtp2")]
```
What is this here for?

Rosme replied:
This is required so that we can access the internals of NGO, and more specifically the `UnityTransport`. We need this to access to the Network Driver, and register callbacks for our adapters to know when the transport is initialized.

JesseOlmer replied:
That's right, I forgot we moved UTP into this assembly from the old one (on line 14 here) :) disregard

JesseOlmer commented on com.unity.netcode.gameobjects/Runtime/Transports/UTP/UnityTransport.cs:
```diff
@@ -1,3 +1,12 @@
+// All references to UNITY_MP_TOOLS_NETSIM_IMPLEMENTATION_ENABLED should be defined in the same way,
+// as any discrepancies are likely to result in build failures
+// ---------------------------------------------------------------------------------------------------------------------
+#if UNITY_EDITOR || ((DEVELOPMENT_BUILD && !UNITY_MP_TOOLS_NETSIM_DISABLED_IN_DEVELOP) || (!DEVELOPMENT_BUILD && UNITY_MP_TOOLS_NETSIM_ENABLED_IN_RELEASE))
```
```suggestion
#if UNITY_EDITOR || (DEVELOPMENT_BUILD && !UNITY_MP_TOOLS_NETSIM_DISABLED_IN_DEVELOP) || (!DEVELOPMENT_BUILD && UNITY_MP_TOOLS_NETSIM_ENABLED_IN_RELEASE)
```

JesseOlmer commented on com.unity.netcode.gameobjects/Runtime/Transports/UTP/UnityTransport.cs:
```diff
@@ -1,3 +1,12 @@
+// NetSim Implementation compilation boilerplate
+// All references to UNITY_MP_TOOLS_NETSIM_IMPLEMENTATION_ENABLED should be defined in the same way,
+// as any discrepancies are likely to result in build failures
+// ---------------------------------------------------------------------------------------------------------------------
```
```suggestion
// as any discrepancies are likely to result in build failures
```

Commit:
Merge branch 'develop' into feature/tools-netsim-support

JesseOlmer commented on com.unity.netcode.gameobjects/Runtime/Transports/UTP/UnityTransport.cs:
```diff
@@ -1,3 +1,12 @@
+// ---------------------------------------------------------------------------------------------------------------------
+#if UNITY_EDITOR || ((DEVELOPMENT_BUILD && !UNITY_MP_TOOLS_NETSIM_DISABLED_IN_DEVELOP) || (!DEVELOPMENT_BUILD && UNITY_MP_TOOLS_NETSIM_ENABLED_IN_RELEASE))
+    #define UNITY_MP_TOOLS_NETSIM_IMPLEMENTATION_ENABLED
+#endif
+// ---------------------------------------------------------------------------------------------------------------------
```
```suggestion
#endif
```

JesseOlmer commented on com.unity.netcode.gameobjects/Runtime/Transports/UTP/UnityTransport.cs:
```diff
@@ -1348,12 +1369,12 @@ private void ConfigureSimulator()
+#endif //MULTIPLAYER_TOOLS_1_0_0_PRE_7
 
-#if UNITY_EDITOR || DEVELOPMENT_BUILD
+#if UNITY_EDITOR || DEVELOPMENT_BUILD || UNITY_MP_TOOLS_NETSIM_IMPLEMENTATION_ENABLED
```
Is UNITY_EDITOR || DEVELOPMENT_BUILD here because the NETSIM define is 2.0 only? Can't we simplify these cases in the first define?

Rosme replied:
We need to maintain the existing Debug Simulator as well if the Tools are not present. But since the Network Simulator can also be included in Release builds if the user wished so, we need all three.

JesseOlmer commented on com.unity.netcode.gameobjects/Runtime/Transports/UTP/UnityTransport.cs:
```diff
@@ -1362,76 +1383,81 @@ private void ConfigureSimulator()
+#endif //MULTIPLAYER_TOOLS_1_0_0_PRE_7
 
 #if UNITY_EDITOR || DEVELOPMENT_BUILD
+#if !UNITY_MP_TOOLS_NETSIM_IMPLEMENTATION_ENABLED
```
This + the one on 1434 feels a bit too messy. Consider extracting methods to simplify this stuff a bit and not doing preprocessor surgery on if/else flow.

Rosme replied:
I'll extract those in methods.

Commit:
Apply suggestions from code review

Co-authored-by: Jesse Olmer <jesseo@unity3d.com>

Commit:
Removal of Debug Simulator in UTP1, updated callback to avoid moving to
much stuff around

Commit:
Merge branch 'develop' into feature/tools-netsim-support

Commit:
Fixed compiling issues

simon-lemay-unity commented:
> Per our alignment, I've updated the PR to remove Debug Simulator if we are using UTP 2. I split the stuff to avoid a mixtures of weird ifdef all over the place.

I understand there's already alignment on this at a higher level than me, but we're okay introducing a breaking API change if someone happens to have UTP 2.0 installed?

Rosme commented:
> > Per our alignment, I've updated the PR to remove Debug Simulator if we are using UTP 2. I split the stuff to avoid a mixtures of weird ifdef all over the place.
> 
> I understand there's already alignment on this at a higher level than me, but we're okay introducing a breaking API change if someone happens to have UTP 2.0 installed?

As we see it, UTP 2 is already considered a breaking change. Technically, you need to get UTP2 to have this breaking change. If you just get the latest of NGO, it's not a breaking change. If you are willing to go to UTP2, you are accepting that change.

simon-lemay-unity commented:
> As we see it, UTP 2 is already considered a breaking change. Technically, you need to get UTP2 to have this breaking change. If you just get the latest of NGO, it's not a breaking change. If you are willing to go to UTP2, you are accepting that change.

I'm not sure I understand what you mean by UTP 2.0 being a breaking change. With the recent changes that landed, someone installing UTP 2.0 will get something that still works without having to modify their code. All breaking API changes introduced by UTP 2.0 are handled by NGO through version defines. But with the proposed changes here, someone could install UTP 2.0 and get a non-compiling project _without any indication as to what happened_.

I understand (and support) the rationale of not wanting to support the existing debug simulator if UTP 2.0 is installed. That's fine. Users with UTP 2.0 should install the tools package to get the new and improved network simulator. But straight out removing the API gives them absolutely no indication that this is what is expected of them. At least change the implementation to throw an exception or log an error or something so that users won't end up with a non-compiling application and no idea of how to proceed. Or at the _very_ least add an entry to the changelog indicating how users are expected to handle this.

becksebenius-unity commented:
> > As we see it, UTP 2 is already considered a breaking change. Technically, you need to get UTP2 to have this breaking change. If you just get the latest of NGO, it's not a breaking change. If you are willing to go to UTP2, you are accepting that change.
> 
> I'm not sure I understand what you mean by UTP 2.0 being a breaking change. With the recent changes that landed, someone installing UTP 2.0 will get something that still works without having to modify their code. All breaking API changes introduced by UTP 2.0 are handled by NGO through version defines. But with the proposed changes here, someone could install UTP 2.0 and get a non-compiling project _without any indication as to what happened_.
> 
> I understand (and support) the rationale of not wanting to support the existing debug simulator if UTP 2.0 is installed. That's fine. Users with UTP 2.0 should install the tools package to get the new and improved network simulator. But straight out removing the API gives them absolutely no indication that this is what is expected of them. At least change the implementation to throw an exception or log an error or something so that users won't end up with a non-compiling application and no idea of how to proceed. Or at the _very_ least add an entry to the changelog indicating how users are expected to handle this.

An alternative to this would to be use `[HideInInspector]` to remove it from the inspector UI, and then use `[Obsolete]` tag to indicate to users that the API should not be used if they have UTP2.0 installed. I believe this would satisfy semantic versioning and give better feedback to users if they're relying on that API.

Rosme commented:
> > > As we see it, UTP 2 is already considered a breaking change. Technically, you need to get UTP2 to have this breaking change. If you just get the latest of NGO, it's not a breaking change. If you are willing to go to UTP2, you are accepting that change.
> > 
> > 
> > I'm not sure I understand what you mean by UTP 2.0 being a breaking change. With the recent changes that landed, someone installing UTP 2.0 will get something that still works without having to modify their code. All breaking API changes introduced by UTP 2.0 are handled by NGO through version defines. But with the proposed changes here, someone could install UTP 2.0 and get a non-compiling project _without any indication as to what happened_.
> > I understand (and support) the rationale of not wanting to support the existing debug simulator if UTP 2.0 is installed. That's fine. Users with UTP 2.0 should install the tools package to get the new and improved network simulator. But straight out removing the API gives them absolutely no indication that this is what is expected of them. At least change the implementation to throw an exception or log an error or something so that users won't end up with a non-compiling application and no idea of how to proceed. Or at the _very_ least add an entry to the changelog indicating how users are expected to handle this.
> 
> An alternative to this would to be use `[HideInInspector]` to remove it from the inspector UI, and then use `[Obsolete]` tag to indicate to users that the API should not be used if they have UTP2.0 installed. I believe this would satisfy semantic versioning and give better feedback to users if they're relying on that API.

I believe the issue that @JesseOlmer was pointing out though is that we still have an existing API that doesn't do anything in that situation, though the `[Obsolete]` part could indicate that.

DenninDalke committed:
Fix compiler error typo: "river" should be "driver"

DenninDalke committed:
Deprecate DebugSimulator when using UTP2

simon-lemay-unity commented on com.unity.netcode.gameobjects/Runtime/Transports/UTP/UnityTransport.cs:
```diff
@@ -648,6 +666,7 @@ public void SetConnectionData(NetworkEndpoint endPoint, NetworkEndpoint listenEn
             SetConnectionData(serverAddress, endPoint.Port, listenAddress);
         }
 
+#if !UTP_TRANSPORT_2_0_ABOVE
```
If only deprecating the debug simulator, should conditionally add an obsolete argument here.

DenninDalke replied:
I'm only deprecating the `DebugSimulator` if we use `UTP_TRANSPORT_2_0_ABOVE` in this case, for UTP1, it's still valid.

simon-lemay-unity replied:
Right, my point is that this piece of code is compiling `SetDebugSimulatorParameters` out entirely if UTP 2.0 is installed. If the intention is to only deprecate it, then we should do as above and conditionally compile the `Obsolete` attribute, and not the entire method.

DenninDalke replied:
I see, that makes sense, I've updated the code.

DenninDalke committed:
Deprecate SetDebugSimulatorParameters on UTP1

DenninDalke committed:
Merge branch 'develop' into feature/tools-netsim-support

JesseOlmer commented on com.unity.netcode.gameobjects/Runtime/Transports/UTP/UnityTransport.cs:
```diff
@@ -352,6 +350,9 @@ public struct SimulatorParameters
         /// - packet drop rate (packet loss)
         /// </summary>
+#if UTP_TRANSPORT_2_0_ABOVE
+        [Obsolete("DebugSimulator is no longer supported and has no effect.")]
```
```suggestion
        [Obsolete("DebugSimulator is no longer supported and has no effect. Use Network Simulator from the Multiplayer Tools package.", false)]
```

Adding reference to Tools package, and also make this a Warning, not an Error.

JesseOlmer commented on com.unity.netcode.gameobjects/Runtime/Transports/UTP/UnityTransport.cs:
```diff
@@ -368,8 +369,8 @@ private struct PacketLossCache
 
-        internal static event Action<UnityTransport> TransportInitialized;
-        internal static event Action<UnityTransport> TransportDisposed;
+        internal static event Action<int, NetworkDriver> TransportInitialized;
+        internal static event Action<int> TransportDisposed;
```
Any reason we can't make these public and then remove the `internalsvisibleto`? I'd love to get rid of "hidden" public apis.

JesseOlmer commented on com.unity.netcode.gameobjects/Runtime/Transports/UTP/UnityTransport.cs:
```diff
@@ -670,6 +671,9 @@ public void SetConnectionData(NetworkEndpoint endPoint, NetworkEndpoint listenEn
         /// <param name="packetJitter">Packet jitter in milliseconds.</param>
         /// <param name="dropRate">Packet drop percentage.</param>
+#if UTP_TRANSPORT_2_0_ABOVE
+        [Obsolete("SetDebugSimulatorParameters is no longer supported and has no effect.")]
```
Same here please, point at tools package. Make warning if it's "safe" to still call this (leave as error if not)

DenninDalke replied:
Done.

JesseOlmer commented on com.unity.netcode.gameobjects/Runtime/Transports/UTP/UnityTransport.cs:
```diff
@@ -1347,24 +1351,34 @@ public override void Shutdown()
-                packetDelayMs: DebugSimulator.PacketDelayMS,
-                packetJitterMs: DebugSimulator.PacketJitterMS,
-                packetDropPercentage: DebugSimulator.PacketDropRate,
+                packetDelayMs: 0,
```
nit: add a comment about these fields being intentionally ignored for future maintainers.

DenninDalke replied:
Done.

DenninDalke committed:
Add tool suggestion to DebugSimulator deprecation

Co-authored-by: Jesse Olmer <jesseo@unity3d.com>

DenninDalke committed:
Add tool suggestion for DebugSimulator deprecation and comment

DenninDalke committed:
Move comments out the parameter list to avoid CI failure

DenninDalke committed:
Fix formatting

DenninDalke committed:
Merge branch 'develop' into feature/tools-netsim-support
